proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,733 @@
+
+1. A qualified name where the last segment is the simple name and the rest is the qualifier.
+2. A dotted simple name.
+
+.. [#] See `Approaches to dotted names`_ for discussion.
+
+Bindings are for simple names only
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Currently, the ``inScope`` relation relates *qualified* names to
+entities, with the following signature: [#]_
+
+.. [#] This signature is taken from [hsmods]_, which like this proposal is mostly concerned
+       with modules. In reality the relation should relate *scopes* to bindings, but we are
+       only care about module scopes at the moment.
+
+::
+
+   inScope :: Rel Module QualName Entity
+
+We propose that instead it relates *simple* names to entities. We continue to use the ``exports``
+relation as it is.
+
+We add a new relation ``refersTo`` which relates qualified names to
+entities in a scope. [#]_
+
+.. [#] By analogy to programming languages: the bindings in scope are like an environment,
+   simple names are like variables, qualified names are like expressions, and ``refersTo``
+   is ``evalExpr``.
+
+::
+
+   inScope :: Rel Module Name Entity
+   exports :: Rel Module Name Entity
+   refersTo :: Rel Module QualName Entity
+
+``refersTo`` works as follows:
+
+- If the qualified name is a (possibly dotted) simple name, then 
+  it refers to whatever that simple name is bound to in the current scope.
+- If the qualified name has a qualifier, then we work out what
+  the qualifier refers to in the current scope. If it refers to a module,
+  then we work out what the simple name refers to in the exported environment of
+  that module.
+
+That is, we interpret ``A.b`` by first working out what ``A`` refers to, as a
+module, and then looking inside it to find ``b``. ``A`` may itself be
+qualified (since modules may export module bindings, just like any other
+bindings), hence the need for the recursive lookup.
+
+Imports
+^^^^^^^
+
+Normal imports can only appear at the top level of a file. The target of a 
+normal import is a *dotted* name, which is resolved using the
+current rules for locating external Haskell modules.
+
+Importing a module brings all the exported bindings of the target module
+into scope, as described in [hsmods]_. It also brings
+the target module into scope bound to its declared simple name (which may be
+a dotted name if it is a file module).
+
+If a module is imported ``qualified`` then the exported bindings of the target
+module are not brought into scope.
+
+If a module is imported ``as`` a simple name, then the target module is brought
+into scope bound to that simple name instead.
+
+Module export specifiers
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The behaviour of the ``module`` export specifiers are changed as follows.
+
+- ``module M`` *changes* to exports the name ``M`` as a module binding.
+- ``module M(a, b, .. c)`` exports ``M`` and the listed in-scope bindings.
+- ``module M(..)`` exports ``M`` and all of its exported bindings. [#]_
```

Comment:
What if `M` is the current module? Then this definition is badly recursive.

Reply:
Good question! I think this is in fact fine with the right kind of recursive implementation. We can get the same kind of issue with recursive modules:
```
module A
    import B
    x = 1

module B
   import A
```

(note the lack of export specifiers, you could also give them `B(..)` and `A(..)` respectively.)
My implementation will say that both `A` and `B` export `x`. In the same vein, `M(..)` when `M` is the current module will just have no effect (it simply never tells us any new `exports` facts).